{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "adjacent-trial",
   "metadata": {},
   "source": [
    "## USING GFF TO CONVERT V5.3 TO V5.6"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "compliant-relations",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/balogu15/.conda/envs/personal/lib/python3.8/site-packages/pandas/core/series.py:4509: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  return super().replace(\n",
      "<ipython-input-31-b2e2b0779124>:10: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  v5_3_mRNA['attributes'] = v5_3_mRNA['attributes'].str.split(';')\n",
      "<ipython-input-31-b2e2b0779124>:11: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  v5_3_mRNA['Name'] = 'nan'\n",
      "/home/balogu15/.conda/envs/personal/lib/python3.8/site-packages/pandas/core/frame.py:4524: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  return super().replace(\n",
      "<ipython-input-31-b2e2b0779124>:18: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  v5_3_mRNA.dropna(axis = 0, subset = ['Name'], inplace = True)\n",
      "/home/balogu15/.conda/envs/personal/lib/python3.8/site-packages/pandas/core/series.py:4509: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  return super().replace(\n",
      "<ipython-input-31-b2e2b0779124>:26: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  v5_6_mRNA['attributes'] = v5_6_mRNA['attributes'].str.split(';')\n",
      "<ipython-input-31-b2e2b0779124>:27: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  v5_6_mRNA['Name'] = 'nan'\n",
      "/home/balogu15/.conda/envs/personal/lib/python3.8/site-packages/pandas/core/frame.py:4524: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  return super().replace(\n",
      "<ipython-input-31-b2e2b0779124>:34: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  v5_6_mRNA.dropna(axis = 0, subset = ['Name'], inplace = True)\n",
      "/home/balogu15/.conda/envs/personal/lib/python3.8/site-packages/pandas/core/series.py:4509: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  return super().replace(\n",
      "/home/balogu15/.conda/envs/personal/lib/python3.8/site-packages/pandas/core/indexing.py:1599: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  self.obj[key] = infer_fill_value(value)\n",
      "/home/balogu15/.conda/envs/personal/lib/python3.8/site-packages/pandas/core/indexing.py:1720: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  self._setitem_single_column(loc, value, pi)\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import gffpandas.gffpandas as gffpd\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "\n",
    "#### ISOLATING THE GENE IN GFF v5.3 ####\n",
    "v5_3 = gffpd.read_gff3('/research/projects/chlamydomonas/MAexpression/analysis/flo/flo/chlamy.5.3.w_organelles_mtMinus_transcripts.gff')\n",
    "v5_3 = v5_3.df\n",
    "v5_3_mRNA = v5_3.loc[v5_3['type'] == 'mRNA']\n",
    "v5_3_mRNA['attributes'].replace('geneName=', 'gene_name=', regex = True, inplace = True)\n",
    "v5_3_mRNA['attributes'] = v5_3_mRNA['attributes'].str.split(';')\n",
    "v5_3_mRNA['Name'] = 'nan'\n",
    "\n",
    "for i in list(v5_3_mRNA.index.values):\n",
    "    for a in v5_3_mRNA.at[i, 'attributes']:\n",
    "        if 'Name=' in a:\n",
    "            v5_3_mRNA.at[i, 'Name'] = a\n",
    "v5_3_mRNA.replace('nan', np.nan, inplace = True)\n",
    "v5_3_mRNA.dropna(axis = 0, subset = ['Name'], inplace = True)\n",
    "v5_3_mRNA['Name'].replace('Name=', '', regex = True, inplace = True)\n",
    "\n",
    "#### ISOLATING THE GENE IN GFF v5.6 ####\n",
    "v5_6 = gffpd.read_gff3('/research/projects/chlamydomonas/MAexpression/analysis/flo/flo/Creinhardtii_281_v5.6.exons.gff3')\n",
    "v5_6 = v5_6.df\n",
    "v5_6_mRNA = v5_6.loc[v5_6['type'] == 'mRNA']\n",
    "v5_6_mRNA['attributes'].replace('geneName=', 'gene_name=', regex = True, inplace = True)\n",
    "v5_6_mRNA['attributes'] = v5_6_mRNA['attributes'].str.split(';')\n",
    "v5_6_mRNA['Name'] = 'nan'\n",
    "\n",
    "for i in list(v5_6_mRNA.index.values):\n",
    "    for a in v5_6_mRNA.at[i, 'attributes']:\n",
    "        if 'Name=' in a:\n",
    "            v5_6_mRNA.at[i, 'Name'] = a\n",
    "v5_6_mRNA.replace('nan', np.nan, inplace = True)\n",
    "v5_6_mRNA.dropna(axis = 0, subset = ['Name'], inplace = True)\n",
    "v5_6_mRNA['Name'].replace('Name=', '', regex = True, inplace = True)\n",
    "\n",
    "for i in list(v5_6_mRNA.index.values):\n",
    "    ### 90 percentile of gene length in v5.6 ###\n",
    "    v5_6_length = v5_6_mRNA.at[i, 'end'] - v5_6_mRNA.at[i, 'start']\n",
    "    percentile = v5_6_length * 0.9\n",
    "    ### identifying the corresponding sections of v5.3 in v6 ###\n",
    "    section = v5_3_mRNA.loc[v5_3_mRNA['seq_id'] == v5_6_mRNA.at[i, 'seq_id']]\n",
    "    section = section.loc[section['strand'] == v5_6_mRNA.at[i, 'strand']]\n",
    "    for a in list(section.index.values):\n",
    "        if v5_6_mRNA.at[i, 'start'] <= section.at[a, 'start'] and section.at[a, 'start'] <= v5_6_mRNA.at[i, 'end'] and v5_6_mRNA.at[i, 'start'] <= section.at[a, 'end'] and section.at[a, 'end'] <= v5_6_mRNA.at[i, 'end']:\n",
    "            v5_6_mRNA.at[i, 'v5.3'] = section.at[a, 'Name']\n",
    "        if v5_6_mRNA.at[i, 'start'] <= section.at[a, 'start'] and (v5_6_mRNA.at[i, 'end'] - section.at[a, 'start']) >= percentile:\n",
    "            v5_6_mRNA.at[i, 'v5.3'] = section.at[a, 'Name']\n",
    "        if v5_6_mRNA.at[i, 'end'] >= section.at[a, 'end'] and (section.at[a, 'end'] - v5_6_mRNA.at[i, 'start']) >= percentile:\n",
    "            v5_6_mRNA.at[i, 'v5.3'] = section.at[a, 'Name']\n",
    "\n",
    "v5_6_mRNA['v5_6_gene'] = v5_6_mRNA['Name'].replace('.t[1-9].*','', regex = True)\n",
    "v5_6_mRNA.to_csv('/research/projects/chlamydomonas/MAexpression/data/genome_info/v5.3_to_v5.6.csv', sep = '\\t', index = False, header = True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "abandoned-audit",
   "metadata": {},
   "source": [
    "## USING GFF TO CONVERT V5.6 TO V6"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "id": "naval-solomon",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "conversion = pd.read_csv('/research/projects/chlamydomonas/MAexpression/data/genome_info/v5_to_v6_liftover/ChlamydomonasTranscriptNameConversionBetweenReleases.Mch12b.csv', delimiter = ',', skiprows = 1)\n",
    "conversion.columns = ['5.5', '3', '1', 'Genbank', '4', '4.3', 'u5', 'u9', '5.3.1']\n",
    "conversion = conversion[~conversion.index.duplicated(keep = 'first')]\n",
    "conversion.drop(['3', '1', 'Genbank', '4', '4.3', 'u5', 'u9'], axis = 1, inplace = True)\n",
    "conversion = conversion.replace(' ', '', regex = True)\n",
    "conversion.set_index('5.3.1', inplace = True)\n",
    "divergence = pd.read_csv('/research/projects/chlamydomonas/MAexpression/data/genome_info/divergence_info/divergence.out.txt', delimiter = '\\t', index_col = 'transcript_ID')\n",
    "PACid2geneID = pd.read_csv('/research/references/chlamydomonas/5.3_chlamy_w_organelles_mt_minus/annotation/PACid2geneID.txt', delimiter = '\\t')\n",
    "PACid2geneID.columns = ['transcript_ID', 'geneid']\n",
    "PACid2geneID.set_index('transcript_ID', inplace = True)\n",
    "divergence = divergence.join(PACid2geneID)\n",
    "divergence.rename(columns = {'geneid': 'v5.3.1'}, inplace = True)\n",
    "divergence.set_index('v5.3.1', inplace = True)\n",
    "divergence = divergence.join(conversion)\n",
    "geneID = pd.read_csv('/research/projects/chlamydomonas/MAexpression/data/genome_info/v5_to_v6_liftover/preliminary_v6_Cre_liftover.tsv', delimiter = '\\t', header = None)\n",
    "geneID.columns = ['v6', 'v5']\n",
    "geneID.set_index('v5', inplace = True)\n",
    "geneID = geneID[~geneID.index.duplicated(keep = 'first')]\n",
    "divergence = divergence.reset_index()\n",
    "divergence['v5_5_gene'] = divergence['5.5'].replace('.t[1-9].*','', regex = True)\n",
    "divergence.set_index('v5_5_gene', inplace = True)\n",
    "divergence = divergence[~divergence.index.duplicated(keep = 'first')]\n",
    "divergence = pd.concat([divergence, geneID], axis = 1)\n",
    "divergence = divergence.reset_index()\n",
    "divergence.rename(columns = {'level_0':'v5_5_gene', 'index':'v5.3.1', '5.5':'JGI_v5.5_ID', 'v6':'JGI_v6'}, inplace = True)\n",
    "divergence.to_csv('/research/projects/chlamydomonas/MAexpression/analysis/divergence/files/divergence_file_v5.3_and_v6.csv', sep = '\\t', header = True, index = True)\n",
    "chlamy_v5_conversion = divergence[['v5_5_gene', 'v5.3.1', 'JGI_v5.5_ID', 'JGI_v6']]\n",
    "chlamy_v5_conversion.to_csv('/research/projects/chlamydomonas/MAexpression/data/genome_info/v5.3_to_v5.6_to_v6.csv', sep = '\\t', header = True, index = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "satellite-stanley",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:.conda-personal] *",
   "language": "python",
   "name": "conda-env-.conda-personal-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
